File: tools/build_doubts.mjs
/**
 * 疑点清单，按「**这是谁的问题**」分。
 *
 * 之前混了两件根本不同的事，用户指出来了：
 *
 *   谱自己留空　　—— **不是问题**。「纪其所可知，阙其所未知」。
 *                     谱没意见，我们也没意见，照样空着。修史本来就不能尽善尽美。
 *   我们读不出来　—— **我们的问题**，该修。
 *   谱写了但两个人都对得上 —— **真分不清**，唯一需要人去认的。
 *
 * 所以第一类根本不该叫「疑点」。它叫「谱上就是这么记的」。
 */
import { readFileSync, writeFileSync } from 'node:fs';
import { buildIndex } from '../src/core/lineage.ts';
import { withBacklinks } from '../src/core/backlink.ts';
import { EraChart } from '../src/core/years.ts';
import { buildWindows, windowNote } from '../src/core/activity.ts';
import { candidates, kept } from '../src/core/candidates.ts';

const J = n => JSON.parse(readFileSync(`data/${n}.json`, 'utf8'));
const people = withBacklinks(J('people'));
const idx = buildIndex(people);
const chart = new EraChart(J('erachart'));
const win = buildWindows(people, chart);
const NS = s => (s || '').replace(/[\s　]/g, '');

// 谱自己写下的「这里没有记录」——那是编谱人的明确表述，不是缺失
const SAID_NONE = ['缺', '未详', '失考', '无考', '不详', '未考', '无记', '阙'];
const saidNone = t => SAID_NONE.some(w => NS(t).includes(w));
// 谱上留了空：「嘉庆　年」「年」——格子在，字没填
const leftBlank = t => /^[一-鿿]{0,4}年?$/.test(NS(t));

const out = {
  分不清: [],        // 谱写了，但两个人都对得上 —— 真要人认的
  读不出: [],        // 谱写全了，是我们没读出来 —— 我们的问题
  谱上留空: [],      // 谱自己没写 / 写了「缺」「未详」—— 不是问题
  谱上对不上: [],    // 谱自己两处数字打架（殁在生之前、年代框成空区间）
  超出范围: [],      // 谱的编纂范围之外（胜二的父亲千五在江西）
  名目对不上: [],
};

for (const p of people) {
  // ── 父亲 ──────────────────────────────────────────
  if (p.parent_edges.length) {
    const cs = candidates(idx, p, chart, win);
    const good = kept(cs);
    const byKind = {};
    for (const c of good) (byKind[c.edge.kind] ??= []).push(c);
    let worst = null;
    for (const [k, v] of Object.entries(byKind)) {
      const uniq = new Set(v.map(c => c.edge.parent));
      if (uniq.size > 1 && (!worst || uniq.size > new Set(worst[1].map(c => c.edge.parent)).size)) {
        worst = [k, v];
      }
    }
    if (worst) {
      const cands = worst[1].map(c => ({
        pid: c.edge.parent, name: c.person?.name ?? c.edge.parent_name,
        src_human: c.person?.src_human ?? '', layout: c.layoutNote,
        printedAbove: c.printedAbove,
        window: windowNote(win.get(c.edge.parent)),
      }));
      const nAbove = cands.filter(c => c.printedAbove).length;
      out.分不清.push({
        pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
        father_name: p.father_name, filiation: p.filiation, kind: worst[0],
        window: windowNote(win.get(p.pid)),
        // 谱把其中一个印在他正上方那一格——翻开那页就看见了
        settled: nAbove === 1,
        cands,
      });
    }
  } else if (p.gen === 1) {
    out.超出范围.push({
      pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
      note: `谱上写父名「${p.father_name}」。他在江西乐平——`
          + '这部谱从迁梅始祖胜二公算起，本来就不收他。这不是断链，是谱的编纂范围。',
    });
  } else if (p.father_name) {
    out.读不出.push({
      pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
      what: '父亲', detail: `谱上写父名「${p.father_name}」${p.filiation}`,
      why: '谱里找不到他单独的一条', src: p.father_src,
    });
  } else if (saidNone(p.raw_text) || NS(p.raw_text).length < 14) {
    out.谱上留空.push({
      pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
      what: '父亲', why: saidNone(p.raw_text) ? '谱自己写了「缺」「未详」' : '谱上这一条本来就只留了名字',
      window: windowNote(win.get(p.pid)),
    });
  } else {
    out.谱上留空.push({
      pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
      what: '父亲', why: '这一条写得挺全，但通篇没提父亲',
      window: windowNote(win.get(p.pid)),
    });
  }

  // ── 谱自己两处对不上 ──────────────────────────────
  const w0 = win.get(p.pid);
  if (w0?.conflict) {
    out.谱上对不上.push({
      pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
      what: '年代', why: w0.conflict,
      birth: p.birth?.text ?? null, death: p.death?.text ?? null,
      age: p.age?.text ?? null,
    });
  }

  // ── 生年 ──────────────────────────────────────────
  const bt = p.birth?.text;
  if (bt && chart.lookup(bt).ad == null) {
    const w = windowNote(win.get(p.pid));
    const rec = {
      pid: p.pid, name: p.name, gen: p.gen, src_human: p.src_human,
      what: '生年', detail: `生「${NS(bt)}」`, why: chart.lookup(bt).why, window: w,
    };
    if (saidNone(bt)) { rec.why = '谱自己写了「缺」'; out.谱上留空.push(rec); }
    else if (leftBlank(bt)) { rec.why = '谱上留了空，没填'; out.谱上留空.push(rec); }
    else out.读不出.push(rec);
  }
}

for (const r of J('revisions')) {
  for (const m of r.members) {
    if (m.pid) continue;
    out.名目对不上.push({
      era: r.era, raw: m.raw, name: m.name ?? null,
      why: m.match ?? '没找到', cands: m.candidates ?? [],
    });
  }
}

writeFileSync('data/doubts.json', JSON.stringify(out, null, 1), 'utf8');
for (const [k, v] of Object.entries(out)) console.log(`  ${String(v.length).padStart(5)}　${k}`);
const s = out.分不清.filter(x => x.settled).length;
console.log(`\n  分不清的 ${out.分不清.length} 条里，**${s} 条谱把其中一个印在本人正上方那一格**`
  + `（翻开那页就看见），剩 ${out.分不清.length - s} 条版式也说不出。`);

File: tools/build_doubts.py
"""
疑点清单：全谱所有「谱上没说清」的地方，一次列全。

**这是这个项目最该有的一页。** 前人修谱的准则写在历代序里：

    不缘情而增，不故意而减
    纪其所可知，阙其所未知

「阙其所未知」——不知道的就空着。那就得有个地方把「空着的」都列出来，
不然「阙」就变成了「看不见」。

分五类，每一类都说清**是什么没说清**：

  A 同名分不清    父名同名多个，用谱自己的规矩（世次、生年、生子名单）都排不掉
  B 往上断了      谱上写了父名，但谱里没有他单独的一条
  C 谱上没写父名  连父名都没有
  D 生卒查不出    有生卒原文，但《甲子録》里查不到那个写法
  E 修谱名目对不上 历届名目里的人，在世系里找不到 / 找到多个

只统计和列出，**不改任何数据、不做任何判断**。
"""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

sys.path.insert(0, "tools")
from check_homonym_years import year_of  # noqa: E402

NS = lambda s: "".join((s or "").split()).replace("　", "")
PID = re.compile(r"^P-(册\d+)-(\d+)-(\d+)-(\d+)-(\d+)$")


def coord(pid):
    m = PID.match(pid)
    return (m.group(1), int(m.group(2)), int(m.group(3))) if m else None


def layout_note(child, father):
    a, b = coord(child["pid"]), coord(father["pid"])
    if not a or not b:
        return ""
    same = child["src"]["section"] == father["src"]["section"]
    if a[2] - b[2] == 1 and a[0] == b[0] and same and b[1] <= a[1]:
        return "同一页的上一代" if a[1] == b[1] else f"同一房，前面第 {b[1]} 页"
    if a[2] == 1 and b[2] == 5:
        return "接在上一段末代"
    return "同一房，但版面位置对不上" if same else f"在另一房「{father['src']['section']}」"


def main() -> None:
    people = json.loads(Path("data/people.json").read_text(encoding="utf-8"))
    idx = {p["pid"]: p for p in people}
    for p in people:
        p["_y"] = year_of((p.get("birth") or {}).get("text"))[0]

    # 反向匹配（跟 src/core/backlink.ts 一样的规矩），先把能接的接上
    claims: dict[str, list] = {}
    for f in people:
        for son in f.get("sons_claimed") or []:
            claims.setdefault(NS(son), []).append(f)
    back: dict[str, list] = {}
    for p in people:
        if p["parent_edges"] or not p["father_name"]:
            continue
        forms = {NS(p["name"])} | {NS(a["form"]) for a in p["aliases"]}
        got = []
        for form in forms:
            for f in claims.get(form, []):
                if f["pid"] != p["pid"] and f["gen"] is not None \
                        and p["gen"] - f["gen"] == 1 and f not in got:
                    got.append(f)
        if got:
            back[p["pid"]] = got

    out = {"A同名分不清": [], "B往上断了": [], "C没写父名": [],
           "D生年查不出": [], "E修谱名目对不上": []}

    for p in people:
        edges = p["parent_edges"]
        if not edges and p["pid"] in back:
            edges = [{"parent": f["pid"], "parent_name": f["name"], "kind": "生父",
                      "evidence": "claim_named"} for f in back[p["pid"]]]

        if edges:
            # 三条客观排除（跟 candidates.ts 一致）
            named = {}
            for e in edges:
                if e["evidence"] == "claim_named":
                    named[e["kind"]] = named.get(e["kind"], 0) + 1
            keep = []
            for e in edges:
                f = idx.get(e["parent"])
                if not f or f["gen"] is None or p["gen"] - f["gen"] != 1:
                    continue
                if named.get(e["kind"]) == 1 and e["evidence"] != "claim_named":
                    continue
                cy, fy = p["_y"], f["_y"]
                if cy and fy and not (13 <= cy - fy <= 75):
                    continue
                keep.append((e, f))
            by_kind: dict[str, set] = {}
            for e, f in keep:
                by_kind.setdefault(e["kind"], set()).add(e["parent"])
            worst = max((len(v) for v in by_kind.values()), default=0)
            if worst > 1:
                kind = next(k for k, v in by_kind.items() if len(v) == worst)
                out["A同名分不清"].append({
                    "pid": p["pid"], "name": p["name"], "gen": p["gen"],
                    "src_human": p["src_human"], "father_name": p["father_name"],
                    "filiation": p["filiation"], "kind": kind,
                    "birth": (p.get("birth") or {}).get("text"),
                    "cands": [{
                        "pid": f["pid"], "name": f["name"], "src_human": f["src_human"],
                        "kind": e["kind"], "birth": (f.get("birth") or {}).get("text"),
                        "year": f["_y"], "layout": layout_note(p, f),
                        "sons": f.get("sons_claimed") or [],
                    } for e, f in keep if e["kind"] == kind],
                })
        elif p["father_name"]:
            out["B往上断了"].append({
                "pid": p["pid"], "name": p["name"], "gen": p["gen"],
                "src_human": p["src_human"], "father_name": p["father_name"],
                "filiation": p["filiation"], "father_src": p["father_src"],
            })
        elif p["gen"] != 1:
            out["C没写父名"].append({
                "pid": p["pid"], "name": p["name"], "gen": p["gen"],
                "src_human": p["src_human"],
            })

        bt = (p.get("birth") or {}).get("text")
        if bt and p["_y"] is None:
            out["D生年查不出"].append({
                "pid": p["pid"], "name": p["name"], "gen": p["gen"],
                "src_human": p["src_human"], "birth": bt,
                "why": year_of(bt)[1],
            })

    rev = json.loads(Path("data/revisions.json").read_text(encoding="utf-8"))
    for r in rev:
        for m in r["members"]:
            if m.get("pid"):
                continue
            out["E修谱名目对不上"].append({
                "era": r["era"], "raw": m["raw"], "name": m.get("name"),
                "why": m.get("match", "没找到"),
                "cands": m.get("candidates") or [],
            })

    Path("data/doubts.json").write_text(
        json.dumps(out, ensure_ascii=False, indent=1), encoding="utf-8")

    print("疑点清单　data/doubts.json\n")
    for k, v in out.items():
        print(f"  {len(v):>5}　{k}")
    print(f"\n  合计 {sum(len(v) for v in out.values())} 条")

    print("\n── A 同名分不清，头 5 条 ──")
    for x in out["A同名分不清"][:5]:
        print(f"\n第{x['gen']}世 {x['name']}　{x['src_human']}")
        print(f"   谱上写父名「{x['father_name']}」{x['filiation']}，分不出哪个是{x['kind']}")
        for c in x["cands"]:
            print(f"     · {c['name']}　{c['src_human']}　{c['layout']}"
                  + (f"　生{c['year']}" if c["year"] else "　生年不详"))


if __name__ == "__main__":
    main()
